ic-canister/ic-canister-macros/src/derive.rs

```diff
@@ -213,28 +201,38 @@ fn expand_upgrade_methods(
             fn __pre_upgrade() {
                 use ::ic_storage::IcStorage;
 
-                #(#state_gets)*
+                #state_get
 
-                ::ic_cdk::storage::stable_save((
-                    #( #state_borrows)*
-                ))
-                .unwrap();
+                ::ic_storage::stable::write(#state_borrow).unwrap();
             }
 
             #[cfg(all(target_arch = "wasm32", feature = "export_api"))]
             #[export_name = "canister_post_upgrade"]
             fn __post_upgrade() {
                 use ::ic_storage::IcStorage;
 
-                let (#( #field_names,)*) = ::ic_cdk::storage::stable_restore().unwrap();
+                let #name = match ::ic_storage::stable::read::<#field_type::Versioned>() {
+                    Ok(val) => val,
+                    Err(e) => panic!("failed to upgrade: {}", e),
```

Comment:
For debugging purposes, I think it's better to use `::ic_cdk::trap` here

Reply:
Good point